### PR TITLE
[MoE][7/n] Keep 3D tensors through MoE, flatten inside GroupedExperts

### PR DESCRIPTION
**Pull request by acisseJZhong**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #3425
* __->__ #3423

## Summary
Address https://github.com/pytorch/torchtitan/pull/3386#discussion_r3269104636

Move `.view(-1, dim)` from `MoE.forward()` into `GroupedExperts.forward()` / `GptOssGroupedExperts.forward()`, so the router and shared experts see 3D `(bs, slen, dim)` tensors. This avoids `StridedShard` when `.view(-1, dim)` is called on a `Shard(1)` 3D DTensor at the MoE boundary.

- `GroupedExperts.forward` / `GptOssGroupedExperts.forward`: flatten to 2D at entry
- `TokenChoiceTopKRouter`: use `dim=-1` and relative dim ops to handle any number of leading dims
- `moe_sharding.py`: update placements for 3D shapes (`Shard(1)` for slen, `Shard(2)` for feature dim)

## Test
```
NCCL_NVLS_ENABLE=0 python scripts/loss_compare.py . main --baseline-module='llama4' --baseline-config='llama4_debugmodel' --baseline-options="--parallelism.pipeline_parallel_degree 1 --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 2 --training.local_batch_size=1" --baseline-ngpus=4 --test-module='llama4' --test-config='llama4_debugmodel' --test-options="--parallelism.pipeline_parallel_degree 1 --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 2 --training.local_batch_size=1" --test-ngpus=4 --assert-equal --no-seed-checkpoint
```

### CODE REVIEW

**Comment by acisseJZhong** [comment]
@claude review

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -86,6 +86,11 @@ def forward(
         DTensor→local conversion on entry and local→DTensor(Partial) wrapping
         on exit. The forward body operates on plain local tensors.
         """
+        x = x.reshape(-1, x.size(-1))
+        top_scores = top_scores.reshape(-1, top_scores.size(-1))
+        selected_experts_indices = selected_experts_indices.reshape(
+            -1, selected_experts_indices.size(-1)
+        )
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: can we use `view` instead of `reshape`, if things are already contiguous

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -237,10 +243,10 @@ def forward(
             scores_for_choice, k=self.top_k, dim=-1, sorted=False
         )
 
-        # top scores shape (bs*slen, top_k)
+        # top scores shape (*, top_k)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: can we be explicit and use `(bs, slen, top_k)`

File: torchtitan/models/gpt_oss/moe.py [hunk 3]
```diff
@@ -145,6 +145,11 @@ def forward(
         selected_experts_indices: torch.Tensor,
     ) -> torch.Tensor:
         """Dispatch tokens to experts, compute, combine, and scatter_add."""
+        x = x.reshape(-1, x.size(-1))
+        top_scores = top_scores.reshape(-1, top_scores.size(-1))
+        selected_experts_indices = selected_experts_indices.reshape(
+            -1, selected_experts_indices.size(-1)
+        )
```
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/moe.py]: same

File: torchtitan/models/common/moe.py [hunk 4]
```diff
@@ -142,17 +147,18 @@ def _debug_force_load_balance_routing(
         self, scores: torch.Tensor
     ) -> tuple[torch.Tensor, torch.Tensor]:
         """Balanced round-robin expert assignment.
-        Returns (selected_experts_indices [N, K] LongTensor, top_scores [N, K] FloatTensor).
+        Returns (selected_experts_indices [*, K] LongTensor, top_scores [*, K] FloatTensor).
         """
-        n_tokens = scores.size(0)
+        leading_shape = scores.shape[:-1]
+        n_tokens = scores[..., 0].numel()
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: Let's assume the dimension is `(bs, slen, _)` and use them explicitly.

File: torchtitan/models/common/moe.py
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: to update

**Review by tianyu-l — APPROVED** [review]
please address comments

**Comment by acisseJZhong** [comment]
@claude review this PR, be concise.
